### PR TITLE
[Serializer] Simplify the metadata factory extension

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13931 |
| License | MIT |

### CODE REVIEW

File: src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php [hunk 1]
```diff
@@ -115,4 +115,9 @@ private function getClass($value)
 
         return ltrim(is_object($value) ? get_class($value) : $value, '\\');
     }
+
+    protected function createMetadataClass($class)
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php]: Can you add some PHPDoc?

Comment (aitboudad) [inline src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php]: also can you move https://github.com/EXSyst/symfony/blob/2.7/src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php#L67-L69 into this method ?

**Comment by stof** [comment]
What is the use case for replacing the ClassMetadata ?

**Comment by Ener-Getick** [comment]
It's to add some custom fields.
For example, I want to add attributes for class, methods and properties.

File: src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php (hunk 1 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php]: Is that ok ?
   
    /**
     * Creates a metadata class
     *
     * @param string $class
     *
     * @return ClassMetadataInterface
     */
    protected function createMetadataClass($class)
    {
        if (!class_exists($class) && !interface_exists($class)) {
            throw new InvalidArgumentException(sprintf('The class or interface "%s" does not exist.', $class));
        }
    
        return new ClassMetadata($class);
    }

Comment (aitboudad) [inline src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php]: Sorry can you revert to old way as It become more complicated (I was thinking that can allow user to apply custom logic for a class name).

Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php]: I don't know if it's useful. But if you think it is we could add a new protected method.

File: src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php [hunk 2]
```diff
@@ -115,4 +116,16 @@ private function getClass($value)
...
     }
+
+    /**
+     * Creates a metadata class
+     *
+     * @param string $class
+     *
+     * @return ClassMetadataInterface
+     */
+    protected function createMetadataClass($class)
```
Comment (Tobion) [inline src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php]: protected come before private methods

Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Mapping/Factory/ClassMetadataFactory.php]: Sorry i didn't know that.
I made a new commit.

**Comment by dunglas** [comment]
:+1:

**Comment by Tobion** [comment]
the phpdoc of the class is now wrong which says `Returns a {@link ClassMetadata}.`. It's not for ClassMetadata only anymore. And this shows a strange programming paradigm: using class inheritance for factories. Doesn't sound right to me. Normally one uses decoraters for that.

**Comment by Ener-Getick** [comment]
We could add a new root `MetadataInterface` for phpdoc ?

**Comment by dunglas** [comment]
@xabbuh btw, using a decorator for the cache (instead of the second parameter of the constructor) would be cleaner. But the validator metadatafactory and probably some others are built the same way.

**Comment by dunglas** [comment]
Ping @symfony/deciders. Not sure of what to do about this one. Not a big fan of this extension point but it can be useful in some cases.

**Comment by stof** [comment]
I'm not a fan of this extension point either.

@Ener-Getick can you give more details about your use case, which could allow us to help you find the right way to extend Symfony to support it ?

**Comment by Ener-Getick** [comment]
I would like to add some new fields in the metadata class to describe the
class methods and properties.

**Comment by nicolas-grekas** [comment]
We could allow changing the name of the instantiated class by adding a new optional argument to the constructor `$className = 'Symfony\Component\Serializer\Mapping\ClassMetadata'`. Would it be enough @Ener-Getick and make more sense @Tobion and @stof ?

**Comment by Ener-Getick** [comment]
Yes, it's exactly what i would like.
I'm just not sure that's cleaner but it works too.

**Comment by dunglas** [comment]
I personally prefer @nicolas-grekas approach.

**Comment by Tobion** [comment]
That woulnd't even support custom constructor arguments.

**Comment by nicolas-grekas** [comment]
@Tobion true, but that still would be enough to add some open-closed powerness.

**Comment by Ener-Getick** [comment]
Is that what you thought ?
Sorry I didn't have the time to take care of this issue...

**Comment by Ener-Getick** [comment]
And I benefit of this thread to ask you a question : do you want to make a powerfull serializer in the symfony package like JMSSerializer or just a simple and easy to use serializer ?

**Comment by mvrhov** [comment]
I'd vote for a second. As The JMSSerializer is slowly becoming unmaintained.

**Comment by fabpot** [comment]
Making a more powerful serializer in core is indeed the goal, and @dunglas is working hard on achieving this (2.7 is already full of great changes in this direction).

**Comment by Ener-Getick** [comment]
Is that possible to help him or does he develop alone ?

And I think that this pull request lost of its utility if a powerful serializer is developed, should I close it ?

**Comment by dunglas** [comment]
Help is very welcome @Ener-Getick! We just need to take a decision about this PR.

**Comment by Ener-Getick** [comment]
If you work on the serializer, I think it may be better to add this PR (or something like) later to not make BC breaks.

**Pull request closed without merging**